File: algorithms/2_medium/M/01557_minimum_number_of_vertices_to_reach_all_nodes/solution.cpp
class Solution {
public:
    static vector<int> findSmallestSetOfVertices(const int n, const vector<vector<int>>& edges) {
        assert(n >= 2 && n <= 100'000);
        auto has_parent = vector<bool>(n, false);

        for (const auto& e : edges) {
            const int u = e[0];
            const int v = e[1];
            assert(u >= 0 && u < n && v >= 0 && v < n);
            has_parent[v] = true;
        }

        auto res = vector<int>();
        for (int i = 0; i < n; i += 1) {
            if (!has_parent[i]) {
                res.push_back(i);
            }
        }

        return res;
    }
};

File: algorithms/2_medium/M/01557_minimum_number_of_vertices_to_reach_all_nodes/solution.rs
impl Solution {
    pub fn find_smallest_set_of_vertices(n: i32, edges: Vec<Vec<i32>>) -> Vec<i32> {
        assert!(n >= 2 && n <= 100_000);
        let mut has_parent = vec![false; n as usize];

        for e in edges {
            let (u, v) = (e[0], e[1]);
            assert!(u >= 0 && u < n && v >= 0 && v < n);
            has_parent[v as usize] = true;
        }

        let mut res = Vec::<i32>::new();
        for i in 0..n {
            if !has_parent[i as usize] {
                res.push(i);
            }
        }

        res
    }
}
